File: PostalCode/ViewController/BaseViewController/BaseTabBarController.m
//
//  BaseTabBarController.m
//  PostalCode
//
//  Created by OCHIISHI Koichiro on 2013/10/22.
//  Copyright (c) 2013年 OCHIISHI Koichiro. All rights reserved.
//

#import "BaseTabBarController.h"
#import "PostalCode-Swift.h"

@implementation BaseTabBarController

- (void)viewDidLoad
{
    [super viewDidLoad];
	// Do any additional setup after loading the view.
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(handleSearchQuery:)
                                                 name:@"handleSearchQuery"
                                               object:nil];
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self
                                                    name:@"handleSearchQuery"
                                                  object:nil];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - SearchViewController

- (void)handleSearchQuery:(NSNotification *)notification
{
    self.selectedIndex = 1;
    BaseNavigationController *navigationController = (BaseNavigationController *)self.viewControllers[1];
    [navigationController popToRootViewControllerAnimated:NO];
    [self performSelector:@selector(afterDelayHandleSearchQuery:) withObject:notification.object[@"query"] afterDelay:0.f];
}

- (void)afterDelayHandleSearchQuery:(NSString *)query
{
    BaseNavigationController *navigationController = (BaseNavigationController *)self.viewControllers[1];
    SearchViewController *viewController = [navigationController.viewControllers firstObject];
    [viewController searchQuery:query];
}

@end

File: PostalCode/ViewController/BaseViewController/BaseTabBarController.swift
//
//  BaseTabBarController.swift
//  PostalCode
//
//  Created by Koichiro OCHIISHI on 2025/05/05.
//  Copyright © 2025 OCHIISHI Koichiro. All rights reserved.
//

import UIKit

@objcMembers
class BaseTabBarController: UITabBarController {

    override func viewDidLoad() {
        super.viewDidLoad()

        NotificationCenter.default.addObserver(self,
                                               selector: #selector(handleSearchQuery(_:)),
                                               name: NSNotification.Name("handleSearchQuery"),
                                               object: nil)
    }

    deinit {
        NotificationCenter.default.removeObserver(self,
                                                  name: NSNotification.Name("handleSearchQuery"),
                                                  object: nil)
    }

    // MARK: - SearchViewController

    @objc private func handleSearchQuery(_ notification: Notification) {
        self.selectedIndex = 1

        guard let navigationController = self.viewControllers?[self.selectedIndex] as? BaseNavigationController else {
            return
        }

        navigationController.popToRootViewController(animated: false)

        if let query = (notification.object as? [String: Any])?["query"] as? String {
            perform(#selector(afterDelayHandleSearchQuery(_:)), with: query, afterDelay: 0.0)
        }
    }

    @objc private func afterDelayHandleSearchQuery(_ query: String) {
        guard let navigationController = self.viewControllers?[1] as? BaseNavigationController,
              let searchViewController = navigationController.viewControllers.first as? SearchViewController else {
            return
        }

        searchViewController.searchQuery(query)
    }
}
